### ISSUE TITLE
Update scipy version

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#549]
Currently we are using scipy 0.17 from 2016 because [later versions could include f90](https://github.com/scipy/scipy/issues/2829#issuecomment-223764054) and we don't know how to compile it.

However, I quickly looked at scipy git log, and the mean edit frequency of for fortran files in scipy is 6 years, so even if f90 is now accepted now, there was no active modernization effort from f77. Which means that later scipy versions could potentially be built, and we could update it. Latest version would be ideal so that we could contribute fixes upstream if needed.

Though the problem is that until dynamic linking is not resolved https://github.com/iodide-project/pyodide/issues/240, if newer versions add additional modules that use LAPACK it will increase the binary size which is not desirable.

**Comment by rth** [comment]
The current status on this is that it's possible to update scipy, but later versions use BLAS/LAPACK in a larger number of modules, which would translate to even bigger package sizes until we address it via https://github.com/iodide-project/pyodide/issues/240 or https://github.com/iodide-project/pyodide/issues/874

**Comment by SaadNaeem96** [comment]
Any updates on when scipy will be updated? Looking to run a function (https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.cossin.html) which is available from scipy 1.5.0

**Comment by 124bit** [comment]
Same thing. Lacking functions from new scipy)))))))

**Comment by certik** [comment]
To move this forward, we are working on compiling SciPy with LFortran: https://gitlab.com/lfortran/lfortran/-/issues/582. We can parse most of the files, some require either minor modifications or (preferably) a few fixes in our fixed-form parser.

I figured the best way would be to get some simplest self-contained part of SciPy working first. It seems the simplest part might be `stats/statlib`: https://gitlab.com/lfortran/lfortran/-/issues/581. It looks like we can parse those files and we are now working on the semantics. It looks like we are not far off, but we still have to implement a few things to get all semantics working. Once that works, we'll see what needs to be fixed in our LLVM backend, probably a few things also.

Once we can generate a correct LLVM IR code for those subroutines, what would be the next step? I assume we select the Wasm backend in the LLVM to generate Wasm. LFortran's runtime library is mostly written in Fortran, so that should translate nicely. There are some things that reach into C (such as file IO), but I think those are not needed for SciPy.

What about things like math functions such as `sin(x)`? We have two versions: one in pure Fortran, which is performance first, the accuracy is roughly 1e-15 relative accuracy and the argument range is roughly `|x| < 1e10`. If `x` is outside that range, it will return incorrect results. These functions are meant for Release mode, when speed is needed. Then we have accuracy first version which are much slower, but those simply call into `libc`, so depend on C. I assume we can add special handling for Wasm into our LLVM backend to call some accurate `sin(x)` implementation that is available for Wasm from somewhere. What is the best approach here?

This is obviously a longer term project, but we will eventually get there. Important is to start now, get simple things working all the way, and then eventually be able to compile all of SciPy. If there is anybody who wants to help us with this work, please let us know. We are always looking for new contributors.

**Comment by rth** [comment]
Great to hear that @certik ! Having alternative compiler for scipy would be very useful, already on x86_64.

> Once we can generate a correct LLVM IR code for those subroutines, what would be the next step? I assume we select the Wasm backend in the LLVM to generate Wasm.

I assume so but I'm not familiar with LLVM. Looks like @StarGate01 managed to make that part work in https://github.com/pyodide/pyodide/issues/184#issuecomment-616948624

> What about things like math functions such as sin(x)? .. What is the best approach here?

Well the scipy test suite  would ideally need to pass on x86_64 (or have its tolerances adjusted). So that will tell how critical it would be. In Pyodide however we currently still have a number of failures, in particular due to an issue with float underflow/overflow handling https://github.com/pyodide/pyodide/issues/156

> Important is to start now, get simple things working all the way, and then eventually be able to compile all of SciPy.

Absolutely, thanks for starting this discussion!

To give a bit more context. Even once there is a compiler that is able to compile Fortran for WASM (related https://github.com/pyodide/pyodide/issues/184) Pyodide has other challenges for building scipy. Due to  function signature mismatches in dynamic linking that apparently work in non WASM environment but that make WASM crash arbitrarily, we currently [have lots of patches](https://github.com/pyodide/pyodide/tree/main/packages/scipy/patches) to try to work around that thanks to work done by @joemarshall . When that makes sense those would need to be contributed to scipy in the long term. At least that's an issue with scipy + CBLAS + CLAPACK + libf2c. It might not be a problem with scipy + BLAS + LAPACK (in Fortran), not sure. Also ideally we would need to switch to a more optimized BLAS #227  in the future.

Also congrats for the progress made on the lfortran project!

**Comment by rth** [comment]
Also to give a general status update on this issue the latest WIP attempt to update is in https://github.com/pyodide/pyodide/pull/1293 but it requires lots of time and effort to fix all the signature mismatches and port patches.

**Comment by certik** [comment]
@rth, awesome, thanks a lot for the tips. It looks like a good first step for us is to:

* Compile the `stats/statlib` with just the regular CPU LLVM backends (x86, arm) and make sure scipy tests pass.

Although perhaps minor for your efforts, that would be a huge milestone for us. I think that's a necessary milestone for any further work and we are not that far off. After that works, we can then work on two things in parallel:

* Implement more Fortran features to make it work for other SciPy modules also
* Get `stats/statlib` working in the browser by reusing the f2py SciPy interface to it and creating a simple Python project using it. 
 
The last point would ensure all the pieces are working with regards to WebAssembly and Pyodide, and people can start using LFortran for simpler Fortran/Python projects. Then to get SciPy working, "all" we need to do is to implement the rest of the Fortran features that other SciPy modules use.

We are looking for more contributors to help us get over the finish line sooner. For example a huge help would be to iron out the WebAssembly and f2py specific interface issues with regards to LFortran, any potential user can help us with that. We can start this work today on just some simple Fortran code and get it working. If anybody reading this is interested in using Fortran in the browser, please get in touch, I am happy to get you started.
